File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/java/CountUnsetBits.java
import java.io.*;
import java.lang.*;
import java.util.*;
import java.math.*;


class CountUnsetBits {
    
    /*@
    public normal_behavior
      ensures \result == ((x % 4294967296L + 6442450944L) % 4294967296L) - 2147483648L;
    model public static pure \bigint fb_wrap(\bigint x);

        public normal_behavior
          requires x >= 0;
          ensures \result == (x == 0 ? 0 : fb_zeroBits(x / 2) + (x % 2 == 0 ? 1 : 0));
          measured_by x;
        model public static pure \bigint fb_zeroBits(int x);
        public normal_behavior
          ensures \result == (n <= 0 ? 0 : fb_totalZeros(n - 1) + fb_zeroBits(n));
          measured_by n > 0 ? n : 0;
        model public static pure \bigint fb_totalZeros(int n);

        public normal_behavior
          assignable \nothing;
          diverges n == Integer.MAX_VALUE;
          ensures n < Integer.MAX_VALUE;
          ensures \result == fb_wrap(fb_totalZeros(n));
    @*/
    //@ code_java_math
    public static int countUnsetBits(int n) {
        int cnt = 0;
        /*@
        loop_invariant n < Integer.MAX_VALUE ==> 1 <= i && (\bigint)i <= (n < 1 ? 1 : (\bigint)n + 1);
        loop_invariant n < Integer.MAX_VALUE ==> cnt == fb_wrap(fb_totalZeros(i - 1));
        loop_writes i, cnt;
        @*/
        for (int i = 1; i <= n; i++) {
            int temp = i;
            /*@
            loop_invariant n < Integer.MAX_VALUE ==> 0 <= temp <= i;
            loop_invariant n < Integer.MAX_VALUE ==> cnt == fb_wrap(fb_totalZeros(i) - fb_zeroBits(temp));
            loop_writes temp, cnt;
            @*/
            while (temp != 0) {
                if ((temp & 1) == 0) {
                    cnt++;
                }
                temp = temp >> 1;
            }
        }
        return cnt;
    }
}

File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/c/CountUnsetBits.c
#include "java_arrays.h"
#include <stdint.h>

/*@
assigns \nothing;
ensures \result == (int32_t)((integer)left + right);
*/
static int32_t java_add(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left + (uint32_t)right);
}

/*@
logic integer fb_zero_bits(integer x) = x <= 0 ? 0 : fb_zero_bits(x / 2) + (x % 2 == 0 ? 1 : 0);
logic integer fb_total_zeros(integer n) = n <= 0 ? 0 : fb_total_zeros(n - 1) + fb_zero_bits(n);
*/
/*@
terminates n < INT32_MAX;
assigns \nothing;
ensures n < INT32_MAX;
ensures \result == (int32_t)fb_total_zeros(n);
*/
int32_t countUnsetBits(int32_t n) {
    int32_t cnt = 0;
    /*@
    loop invariant n < INT32_MAX ==> 1 <= i <= (n < 1 ? 1 : (integer)n + 1);
    loop invariant n < INT32_MAX ==> cnt == (int32_t)fb_total_zeros(i - 1);
    loop assigns i, cnt;
    */
    for (int32_t i = 1; i <= n; i = java_add(i, 1)) {
        int32_t temp = i;
        /*@
        loop invariant n < INT32_MAX ==> 0 <= temp <= i;
        loop invariant n < INT32_MAX ==> cnt == (int32_t)(fb_total_zeros(i) - fb_zero_bits(temp));
        loop assigns temp, cnt;
        */
        while (temp != 0) {
            if ((temp & 1) == 0) {
                cnt = java_add(cnt, 1);
            }
            temp = temp >> 1;
        }
    }
    return cnt;
}
